Breadcrumb: Update styles based on the Domains breadcrumb

epeicher opened the pull request:
#### Changes proposed in this Pull Request

The Domain breadcrumb styles have been migrated to the generic breadcrumb component used in Plugins page.

**Note:** The behaviour of the existing Breadcrumb component has **not** been
modified. The only difference is that the back icon will not be displayed
when the page is in the root component so there is only one item

Additionally, the component has been refactored to cope with three
different conditions:
1. When it is in the compact view and there are multiple items:
  - Only show Back navigation
2. When it is not in the compact view and there are multiple items:
  - Standard breadcrumb behaviour with navigation for the parents
3. When there is only one item:
  - No back button is displayed

Additionally, the following task has been created to use this generic component in the Domains pages:
- https://github.com/Automattic/wp-calypso/issues/61952

|Before | After|
|-------|------|
|<img width="470" alt="CleanShot 2022-03-15 at 12 20 53@2x" src="https://user-images.githubusercontent.com/3519124/158367193-813ac16b-d277-4c80-b6ac-b62400bd64dc.png">|<img width="526" alt="image" src="https://user-images.githubusercontent.com/3519124/158792390-3611b16b-505a-4a84-8c70-d5b97d6b647c.png">|
|**Domains**|<img width="490" alt="image" src="https://user-images.githubusercontent.com/3519124/158367401-22aeff94-03ae-4dd6-8a21-b465ff96a705.png">|

##### Responsive behaviour
|Before | After|
|-------|------|
|![CleanShot 2022-03-15 at 12 26 38](https://user-images.githubusercontent.com/3519124/158368323-80200c98-efb7-4035-92e8-eaadea8cfd3c.gif)|![CleanShot 2022-03-17 at 11 35 16](https://user-images.githubusercontent.com/3519124/158792497-7da7a750-8e7f-4a8b-9ab3-96d8d2019518.gif)|
|**Domains**|![CleanShot 2022-03-15 at 12 24 31](https://user-images.githubusercontent.com/3519124/158368296-93e4fce5-ef5e-41f2-b045-89aa3643ea4e.gif)|

#### Testing instructions

* Open plugins and select one plugin
* Check the styles and compare them against the breadcrumb in the Domains page. The Domains page can be accessed by navigating to Upgrades -> Domains and selecting one domain.
* Check that the responsive behaviour works as expected resizing the window.

Fixes 
- #61143
- #61917

This PR has been created as a consequence of the comments in the related PR:
- #61851

epeicher committed:
Breadcrumb: Update styles based on the Domains one

*Note:* The behaviour of the existing Breadcrumb component has *not* been
modified. Only difference is that the back icon will not be displayed
when the page is in the root component so there is only one item

Additionally, the component has been refactored to cope with three
different conditions:
1. When it is in the compact view and there are multiple items:
  - Only show Back navigation
2. When it is not in the compact view and there are multiple items:
  - Standard breadcrumb behaviour with navigation for the parents
3. When there is only one item:
  - No back button is displayed

epeicher committed:
Fix e2e test. Remove li selector as not used in Back button

epeicher committed:
Fix a couple of style issues

cpapazoglou commented:
Can we also add two Domains related features:
- the bubble icon
<img width="400" alt="CleanShot 2022-03-17 at 14 07 06@2x" src="https://user-images.githubusercontent.com/12430020/158805216-ab4d247c-a16f-47dd-82f1-8f3cb5ff6067.png">

- ability to set the mobile item name and when `null` use the default `Back`

cpapazoglou commented on client/components/breadcrumb/index.tsx:
```diff
@@ -41,33 +54,38 @@ interface Props {
+		);
+	}
+
+	if ( items.length === 1 ) {
+		return <StyledRootLabel>{ items[ 0 ].label }</StyledRootLabel>;
+	}
```
can we move this to the top?

epeicher replied:
Yep

cpapazoglou commented on client/components/breadcrumb/index.tsx:
```diff
@@ -41,33 +54,38 @@ interface Props {
+		return <StyledRootLabel>{ items[ 0 ].label }</StyledRootLabel>;
+	}
+
+	return null;
```
is this really needed? would it make sense moving this top the top in case `items.length === 0`?

epeicher replied:
Yes, it is needed because otherwise, it would return `undefined` (if `length` is empty) and the current version of React does not support it. As a side note, latest React 18 allows that as [explained here](https://github.com/reactwg/react-18/discussions/75).

I will move it to the top and make it more explicit as you suggest

epeicher replied:
Update on the comment above. Typescript enforces to return at least a value, otherwise, it raises the following warning
```
Type '{ (props: PropsWithChildren<Props>): Element | null | undefined; defaultProps: Partial<Props> | undefined; }' is not assignable to type 'FunctionComponent<Props>'.
  Type 'Element | null | undefined' is not assignable to type 'ReactElement<any, any> | null'.
    Type 'undefined' is not assignable to type 'ReactElement<any, any> | null'.
```
which is a way to say `Undefined cannot be returned`

cpapazoglou commented on client/components/breadcrumb/index.tsx:
```diff
@@ -3,34 +3,47 @@ import styled from '@emotion/styled';
-		--color-link: var( --studio-gray-50 );
+const StyledBackLink = styled.a`
+	${ flexAligned };
+	color: var( --studio-gray-80 ) !important; // It uses --studio-gray-50 if not using important
```
Can we remove the `!important` in favor of

```
&:visited {
		color: var( --studio-gray-80 );
	}
```

epeicher replied:
That will not display the color until the user visits the link, is that what we want here?

gcsecsey replied:
I think we just have to override other pseudo-classes of the `a` element. This should work because the `a` element with the styled components class will have priority based on specificity:
```suggestion
	&,
	&:hover,
	&:focus,
	&:visited {
		color: var( --studio-gray-80 );
	}
```

epeicher committed:
Add the bubble functionality to breadcrumbs

It also adds a bubble to the Plugins root item of the breadcrumb.

epeicher commented:
Thanks @cpapazoglou  for your comments!

Regarding

> * ability to set the mobile item name and when `null` use the default `Back`

Do you mean a new string with the label of the Button? Something like 
`{ mobileItem ? mobileItem : translate( 'Back' ) }`

Or were you thinking of something with more functionality like an object representing a breadcrumb item button? [renderMobileItem](https://github.com/Automattic/wp-calypso/blob/acbc00c7cab519a9e3721aac2f790196c3b470ce/client/my-sites/domains/domain-management/components/breadcrumbs/index.jsx#L97)

epeicher committed:
Add a mobileItem label to Breadcrumb to render in compact mode

If passed, `mobileItem` label will be rendered in compact mode. If not
passed, the default "Back" label will be rendered in compact mode.

This commit also updates the docs to include two more examples

epeicher committed:
Fix warning: React element must return at least one value

The previous change was not correct as all return statements were under
an if condition

cpapazoglou commented:
> Thanks @cpapazoglou for your comments!
> 
> Regarding
> 
> > * ability to set the mobile item name and when `null` use the default `Back`
> 
> Do you mean a new string with the label of the Button? Something like `{ mobileItem ? mobileItem : translate( 'Back' ) }`
> 
> Or were you thinking of something with more functionality like an object representing a breadcrumb item button? [renderMobileItem](https://github.com/Automattic/wp-calypso/blob/acbc00c7cab519a9e3721aac2f790196c3b470ce/client/my-sites/domains/domain-management/components/breadcrumbs/index.jsx#L97)
> 
> I have uploaded a [828366d with my assumption](https://github.com/Automattic/wp-calypso/pull/61995/commits/828366db760bc234583d178d6ac2b6c3df6f004a) but happy to change it if required

perhaps just a string is enough. **If we need** something more advanced in the future, there is a chance that we need a `mobile breadcrumb` state that will allow for more flexibility.

gcsecsey reviewed (approved):
LGTM, looks and works as expected. I've left a comment regarding the styling.

- Styles match Domains breadcrumbs ✅ 
- When it is in the compact view and there are multiple items, only show Back navigation ✅ 
- When it is not in the compact view and there are multiple items, standard breadcrumb behavior with navigation for the parents ✅ 
- When there is only one item, no back button is displayed ✅

epeicher committed:
Remove css important mmodifier and use pseudo-classes instead
